bloom/infra/data_storage.py

```diff
@@ -0,0 +1,229 @@
+import asyncio
+import csv
+import functools
+import re
+from collections.abc import Callable, Coroutine
+from enum import Enum
+from logging import getLogger
+from os import environ
+from pathlib import Path
+from typing import Any
+
+import aiofiles
+from aiobotocore.session import ClientCreatorContext, get_session
+from aiocsv import AsyncDictWriter
+from botocore.exceptions import ClientError, EndpointConnectionError
+from pydantic import BaseModel
+
+logger = getLogger()
+
+
+def async_to_sync(coro: Coroutine) -> Callable:
+    """Convert a coroutine to a function."""
+
+    @functools.wraps(coro)
+    def wrapper(storage: "DataStorage", *args: Any, **kwargs: Any) -> Any:
+        async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
+            async with storage.create_client() as client:
+                # We are not giving client as an argument to avoid changing
+                # methods signature which would make ide information
+                # confusing. Methods that are not wrapped explicitely
+                # are therefore prefixed with __ to "prevent" their use
+                # elsewhere.
+                storage.client = client
+                result = await coro(storage, *args, **kwargs)
+                storage.client = None
+                return result
+
+        loop = asyncio.get_event_loop()
+        return loop.run_until_complete(async_wrapper(*args, **kwargs))
+
+    return wrapper
+
+
+class MissingAWSCredentialsError(Exception):
+    def __init__(self) -> None:
+        super().__init__(
+            "AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID "
+            "environment variables should be defined before running the app.",
+        )
+
+
+class StorageType(Enum):
+    """Type of bucket storage.
+
+    More information about them can be found there:
+    https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-class-intro.html
+    """
+
+    STANDARD = 1
+    REDUCED_REDUNDANCY = 2
+    STANDARD_IA = 3
+    ONEZONE_IA = 4
+    INTELLIGENT_TIERING = 5
+    GLACIER = 6
+    DEEP_ARCHIVE = 7
+    OUTPOSTS = 8
+    GLACIER_IR = 9
+
+
+class DataStorage:
+    """An S3 data storage with methods to append elements and to download csv files.
+
+    It is element agnostic and path almost-agnostic. It has a knowledge
+    on path-building to clean outdated cache. This can be improved later on if needed.
+
+    Elements sould be given as BaseModel so that this class is not dependent of any
+    of them.
+
+    """
+
+    def __init__(self, storage_type: StorageType = StorageType.STANDARD) -> None:
+        if "AWS_ACCESS_KEY_ID" not in environ or "AWS_SECRET_ACCESS_KEY" not in environ:
+            raise MissingAWSCredentialsError()
+
+        self._end_point_url = "https://s3.fr-par.scw.cloud"
```

Comment:
hum we should find a way to manage both dev env & production env. In dev env we could still use the docker filesystem ?

Reply:
yes, je peux ajouter ça.